### PR TITLE
Upgrade teachers to use new metrics

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
Continued work on #2162. This PR upgrades the standard teachers to keep all of their metrics in the new Metrics objects.

This diff mostly just marshals the objects around and offers a small refactor of `metrics.py` to make it less obtuse.  The next PR will begin to start moving agent metrics computations, and improving the aggregations logic.

**Testing steps**
CI.

### CODE REVIEW

File: parlai/core/metrics.py [hunk 1]
```diff
@@ -14,16 +14,16 @@
...
+from parlai.core.message import Message
 from parlai.utils.misc import no_lock, round_sigfigs, warn_once
-from parlai.utils.thread import SharedTable
 from parlai.utils.typing import TScalar
 
 
-DEFAULT_METRICS = {'correct', 'bleu-4', 'accuracy', 'f1'}
+DEFAULT_METRICS = {'bleu-4', 'accuracy', 'f1'}
 ROUGE_METRICS = {'rouge-1', 'rouge-2', 'rouge-L'}
 BLEU_METRICS = {'bleu-1', 'bleu-2', 'bleu-3'}
```
Comment (stephenroller) [inline parlai/core/metrics.py]: add bleu-4 to this.

File: parlai/core/metrics.py [hunk 2]
```diff
@@ -338,45 +358,34 @@ class Metrics(object):
     Class that maintains evaluation metrics over dialog.
     """
 
+    @staticmethod
+    def _infer_metrics(cli_arg: str) -> Set[str]:
+        """
+        Parse the CLI metric into a list of metrics we wish to compute.
+        """
+        col: Set[str] = set()
```
Comment (stephenroller) [inline parlai/core/metrics.py]: should this start out as default metrics? not sure.

**Comment by stephenroller** [comment]
Probably a good time for a review.

File: parlai/utils/typing.py [hunk 3]
```diff
@@ -36,4 +25,12 @@ class _Shared(Dict[str, Any]):
...
 
-TScalar = TypeVar('TScalar', bound=_Scalar)
+TScalar = Union[int, float, torch.Tensor]
+"""
+ParlAI type to represent an object that is theoretically expressible as a scalar value.
+Ints and floats are clearly scalars, and torch.Tensors can be represented by a scalar if
+Tensor.numel() == 1. Used as input type for classes derived from Metric.
+
+Note that _Scalar cannot be defined as a subclass of Union, analogously to _Shared
+below, because Union does not support subclassing.
```
Comment (EricMichaelSmith) [inline parlai/utils/typing.py]: nit: _Shared is above TScalar in the code now

File: parlai/core/metrics.py [hunk 4]
```diff
@@ -115,39 +127,157 @@ class AverageMetric(Metric):
...
 
-    def __init__(self, numer: TScalar, denom: TScalar):
-        self._numer = self.as_float(numer)
-        self._denom = self.as_int(denom)
+    __slots__ = ('_numer', '_denom')
 
-    def __add__(self, other: 'AverageMetric') -> 'AverageMetric':
+    def __init__(self, numer: TScalar = 0, denom: TScalar = 1):
+        self._numer = self.as_number(numer)
+        self._denom = self.as_number(denom)
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Hmm - you want denom to be able to be a float now?

File: parlai/core/metrics.py [hunk 5]
```diff
@@ -78,32 +89,33 @@ def as_number(self, obj: TScalar) -> Union[int, float]:
...
+        else:
+            assert isinstance(sum_, (int, float))
+            self._sum = sum_
 
-    def __add__(self, other: 'SumMetric') -> 'SumMetric':
+    def __add__(self, other: Optional['SumMetric']) -> 'SumMetric':
         # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
         # we drop Python 3.6
-        full_sum: float = self._sum + other._sum
+        if other is None:
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Maybe good to add a unit test for other = None? Not sure how thorough we need to be :)

File: parlai/core/metrics.py [hunk 6]
```diff
@@ -115,39 +127,157 @@ class AverageMetric(Metric):
     Class that keeps a running average of some metric.
     """
 
-    def __init__(self, numer: TScalar, denom: TScalar):
-        self._numer = self.as_float(numer)
-        self._denom = self.as_int(denom)
+    __slots__ = ('_numer', '_denom')
 
-    def __add__(self, other: 'AverageMetric') -> 'AverageMetric':
+    def __init__(self, numer: TScalar = 0, denom: TScalar = 1):
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Hmm - why these values for defaults? I could see a situation in which I forget one of these values and then get an unexpected number as a result without warning

File: parlai/core/metrics.py [hunk 7]
```diff
@@ -385,175 +414,101 @@ def __repr__(self):
...
-                        self.metrics[b] += b_score
-                        self.metrics[f'{b}_cnt'] += 1
-                if 'rouge-L' in self.metrics and rouge1 is not None:
-                    self.metrics['rouge-1'] += rouge1
-                    self.metrics['rouge-1_cnt'] += 1
-                    self.metrics['rouge-2'] += rouge2
-                    self.metrics['rouge-2_cnt'] += 1
-                    self.metrics['rouge-L'] += rougeL
-                    self.metrics['rouge-L_cnt'] += 1
+            self._add('accuracy', ExactMatchMetric.compute(prediction, labels))
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: is it worth adding .compute() as an abstract method to Metric to detail when it's needed?

File: parlai/core/metrics.py (hunk 4 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: No, I want it underspecified.

File: parlai/core/metrics.py (hunk 7 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: I don't think so. This is very much an idiom for these standard metrics and I don't expect it to go elsewhere.

File: parlai/core/metrics.py (hunk 6 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Numer could go without a default. Denom's default i think is very reasonable.

Comment (EricMichaelSmith) [reply parlai/core/metrics.py]: Yeah, up to you, but personally I like the idea of leaving at least one arg without a default for each metric to avoid the inevitable silent error from forgetting to pass in any args at all. For SumMetric this wouldn't change the value at all so maybe it's not necessary, but for AverageMetric it would

File: parlai/core/metrics.py (hunk 5 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: I don't think so. I'm not interested in that case in principal. It's just a convenience factor for the user of the class, and if that became no longer necessary, I wouldn't feel a clear need for that case to be covered.

File: parlai/agents/drqa/utils.py [hunk 8]
```diff
@@ -139,7 +139,7 @@ def batchify(batch, null=0, cuda=False):
     # Batch documents and features
     max_length = max([d.size(0) for d in docs])
     x1 = torch.LongTensor(len(docs), max_length).fill_(null)
-    x1_mask = torch.ByteTensor(len(docs), max_length).fill_(1)
+    x1_mask = torch.BoolTensor(len(docs), max_length).fill_(1)
```
Comment (stephenroller) [inline parlai/agents/drqa/utils.py]: drqa changes were because pytorch 1.4 was throwing a fit when I was trying to debug why it was broken by my changes

File: parlai/scripts/interactive_rank.py [hunk 9]
```diff
@@ -1,106 +0,0 @@
-#!/usr/bin/env python3
```
Comment (stephenroller) [inline parlai/scripts/interactive_rank.py]: Deleted because it wasn't worth updating this.

**Comment by stephenroller** [comment]
```
$ python tests/datatests/test_new_tasks.py
[creating task(s): fvqa:DefaultTeacher]
[building data: /private/home/roller/working/parlai/data/FVQA]
[ downloading: https://dl.dropboxusercontent.com/s/iyz6l7jhbt6jb7q/new_dataset_release.zip to /private/home/roller/working/parlai/data/FVQA/FVQA.zip ]
Downloading FVQA.zip: 100%|████████████████████████████████████| 473M/473M [00:20<00:00, 23.3MB/s]
[ Checksum Successful ]
unpacking FVQA.zip
loading: /private/home/roller/working/parlai/data/FVQA/new_dataset_release/all_qs_dict_release.json
[ loaded 3048 episodes with a total of 3048 examples ]
[creating task(s): fvqa:SplitTeacher]
loading: /private/home/roller/working/parlai/data/FVQA/new_dataset_release/all_qs_dict_release.json
[ loaded 3048 episodes with a total of 3048 examples ]
[creating task(s): self_feeding:AllTeacher]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_hh.txt]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_fb.txt]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_st.txt]
/private/home/roller/working/parlai/parlai/scripts/verify_data.py:55: UserWarning: warning: label mentioned twice in candidate_labels:

  warn_once(txt)
/private/home/roller/working/parlai/parlai/scripts/verify_data.py:55: UserWarning: warning: label missing in candidate_labels:

  warn_once(txt)
[ loaded 193777 episodes with a total of 193777 examples ]
[creating task(s): self_feeding:DiafeeTeacher]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_hh.txt]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_fb.txt]
[ loaded 192777 episodes with a total of 192777 examples ]
[creating task(s): self_feeding:DiasatTeacher]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_hh.txt]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_st.txt]
[ loaded 132438 episodes with a total of 132438 examples ]
[creating task(s): self_feeding:FeedbackTeacher]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_fb.txt]
[ loaded 61339 episodes with a total of 61339 examples ]
[creating task(s): self_feeding:SatisfactionTeacher]
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_st.txt]
[ loaded 1000 episodes with a total of 1000 examples ]
[creating task(s): self_feeding:SelfFeedingMTLTeacher]
Warning: SelfFeedingteacher should be assigned subtask. Defaulting to dialog
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_hh.txt]
[ loaded 131438 episodes with a total of 131438 examples ]
[creating task(s): self_feeding:SelfFeedingTeacher]
Warning: SelfFeedingteacher should be assigned subtask. Defaulting to dialog
[ Loading Self-Feeding text data:/private/home/roller/working/parlai/data/self_feeding/train_hh.txt]
[ loaded 131438 episodes with a total of 131438 examples ]
.
----------------------------------------------------------------------
Ran 1 test in 259.108s

OK
python tests/datatests/test_new_tasks.py  225.25s user 8.94s system 88% cpu 4:23.22 total
```

**Comment by stephenroller** [comment]
Alright enough changed that I want a couple new reviews.

File: parlai/core/metrics.py [hunk 10]
```diff
@@ -57,53 +65,77 @@ class Metric(ABC):
...
+    def __float__(self) -> float:
+        return self.value()
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Metric):
+            return self.value() == other.value()
+        else:
+            return self.value() == other
+
+    def __lt__(self, other: Any) -> bool:
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Do you need to overload greater-than as well?

File: parlai/core/metrics.py [hunk 11]
```diff
@@ -115,39 +147,157 @@ class AverageMetric(Metric):
...
+            return 0, 0, 0
+        precision = 1.0 * num_same / len(pred_items)
+        recall = 1.0 * num_same / len(gold_items)
+        f1 = (2 * precision * recall) / (precision + recall)
+        return precision, recall, f1
+
+    @staticmethod
+    def compute(guess: str, answers: List[str]) -> 'F1Metric':
+        if guess is None or answers is None:
+            return AverageMetric(0, 0)
```
Comment (EricMichaelSmith) [inline parlai/core/metrics.py]: Hmm - this isn't division by zero?

File: parlai/core/teachers.py [hunk 12]
```diff
@@ -16,16 +16,14 @@
...
-     ** NOTE: ** We plan to deprecate this method soon in favor of ParlAIDialogTeacher,
-     however several existing tasks are currently still using it.
-
     ``ParlAIDialogTeacher(DialogTeacher)``
      Teacher class that provides access to data in the ParlAI Dialog format.
      See the class description for more details.
 
+    ``FbDialogTeacher(DialogTeacher)``
+     Teacher class that provides access to data in the Facebook Dialog format.
+     See the class description for more details. **This class is deprecated**.
```
Comment (EricMichaelSmith) [inline parlai/core/teachers.py]: 😍

File: parlai/core/metrics.py (hunk 10 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: no, the [total_ordering](https://docs.python.org/3/library/functools.html#functools.total_ordering) decorator fills out the rest.

File: parlai/core/metrics.py (hunk 11 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Yes, that's fine with me. It's undefined afterall.

File: parlai/core/metrics.py [hunk 13]
```diff
@@ -14,18 +14,25 @@
...
+from parlai.core.message import Message
+from parlai.utils.misc import round_sigfigs, warn_once
 from parlai.utils.typing import TScalar
 
+try:
+    import torch.multiprocessing as multiprocessing
+except ImportError:
+    import multiprocessing  # type: ignore
 
-DEFAULT_METRICS = {'correct', 'bleu-4', 'accuracy', 'f1'}
```
Comment (emilydinan) [inline parlai/core/metrics.py]: what the heck is `correct` lol i've never noticed that before

File: parlai/core/metrics.py [hunk 14]
```diff
@@ -14,18 +14,25 @@
...
+
+DEFAULT_METRICS = {'bleu-4', 'accuracy', 'f1'}
 ROUGE_METRICS = {'rouge-1', 'rouge-2', 'rouge-L'}
-BLEU_METRICS = {'bleu-1', 'bleu-2', 'bleu-3'}
+BLEU_METRICS = {'bleu-1', 'bleu-2', 'bleu-3', 'bleu-4'}
```
Comment (emilydinan) [inline parlai/core/metrics.py]: i guess bleu-4 wasn't included before because it is in default

File: parlai/core/metrics.py [hunk 15]
```diff
@@ -115,39 +147,157 @@ class AverageMetric(Metric):
...
+        precision = 1.0 * num_same / len(pred_items)
+        recall = 1.0 * num_same / len(gold_items)
+        f1 = (2 * precision * recall) / (precision + recall)
+        return precision, recall, f1
+
+    @staticmethod
+    def compute(guess: str, answers: List[str]) -> 'F1Metric':
+        if guess is None or answers is None:
+            return AverageMetric(0, 0)
+        g_tokens = normalize_answer(guess).split()
```
Comment (emilydinan) [inline parlai/core/metrics.py]: stray thought: it seems we always split on spaces after normalizing the answer. can we just make `normalize_answer` return the split list?

File: parlai/core/metrics.py [hunk 16]
```diff
@@ -115,39 +147,157 @@ class AverageMetric(Metric):
...
+            smoothing_function=nltkbleu.SmoothingFunction(epsilon=1e-12).method1,
+            weights=weights,
+        )
+        return BleuMetric(score)
+
 
-    def white_space_fix(text):
-        return ' '.join(text.split())
+class RougeMetric(AverageMetric):
+    _evaluator = None
```
Comment (emilydinan) [inline parlai/core/metrics.py]: what does this do

File: parlai/core/metrics.py [hunk 17]
```diff
@@ -335,225 +391,162 @@ def aggregate_metrics(reporters):
...
-                    self.metrics[k] = 0
-                else:
-                    self.metrics[k] = 0.0
-                self.metrics[k + '_cnt'] = 0
-            for k in self.eval_pr:
-                self.metrics['hits@' + str(k)] = 0
-            self.metrics['hits@_cnt'] = 0
+            for uk, v in observation['metrics'].items():
+                if uk in ALL_METRICS:
+                    # don't let the user override our metrics
```
Comment (emilydinan) [inline parlai/core/metrics.py]: hmmm. interesting

File: parlai/core/metrics.py [hunk 18]
```diff
@@ -335,225 +391,162 @@ def aggregate_metrics(reporters):
...
+                    uk = f'USER_{k}'
```
Comment (emilydinan) [inline parlai/core/metrics.py]: didn't you mean `uk` here?

File: parlai/core/metrics.py (hunk 16 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: Are you asking about the terrible diff formatting, or the `_evaluator`? It keeps the state of the Rouge calculator in memory (which rouges we're computing, the tokenizer, [some other stuff](https://github.com/Diego999/py-rouge/blob/16f225b1f46e9d382f1bf5170546da218ee98003/rouge/rouge.py#L85-L89)... )

File: parlai/core/metrics.py (hunk 18 above)
Comment (stephenroller) [reply parlai/core/metrics.py]: yes thx
